File: frontend-v2/src/types/config-promotion.ts
/**
 * Cross-Cluster Config Promotion types (UX-013).
 *
 * Mirrors backend Pydantic models from routes/config_promotion.py.
 */

/** Action type for a promotion change */
export type PromotionAction = 'add' | 'modify' | 'remove';

/** Single resource change in a promotion diff */
export interface PromotionChange {
  action: PromotionAction;
  kind: string;
  name: string;
  namespace: string;
  diff: string | null;
}

/** Cluster identity summary */
export interface PromotionClusterInfo {
  id: number;
  name: string;
}

/** Request body for POST /api/config/promote */
export interface PromoteRequest {
  source_cluster_id: number;
  target_cluster_id: number;
  dry_run: boolean;
}

/** Apply results (only present when applied=true) */
export interface PromotionApplyResults {
  applied: Array<{ kind: string; name: string; namespace: string }>;
  failed: Array<{ kind: string; name: string; namespace: string; error: string }>;
  skipped: Array<{ kind: string; name: string; namespace: string; reason: string }>;
}

/** Response from POST /api/config/promote */
export interface PromoteResponse {
  source_cluster: PromotionClusterInfo;
  target_cluster: PromotionClusterInfo;
  changes: PromotionChange[];
  total_changes: number;
  applied: boolean;
  apply_results: PromotionApplyResults | null;
}

/** Promotion history entry */
export interface PromotionHistoryEntry {
  id: number;
  source_cluster_name: string;
  target_cluster_name: string;
  change_count: number;
  applied_by: string | null;
  applied_at: string;
  success: boolean;
  applied_count: number;
  failed_count: number;
  skipped_count: number;
  changes_summary: Array<{
    action: PromotionAction;
    kind: string;
    name: string;
    namespace: string;
  }> | null;
}

/** Response from GET /api/config/promotions */
export interface PromotionHistoryResponse {
  total: number;
  items: PromotionHistoryEntry[];
}

File: backend/models/config_promotion.py
"""ConfigPromotion model — promotion history for cross-cluster config promotion (UX-013)."""

from sqlalchemy import JSON, Boolean, Column, DateTime, ForeignKey, Index, Integer, String
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from database import Base


class ConfigPromotion(Base):
    """
    Audit log entry for cross-cluster config promotions.

    Records every promotion (dry_run=false) with source/target cluster IDs,
    change count, who applied it, and when. Provides a complete audit trail
    for cross-cluster configuration synchronization.
    """
    __tablename__ = "config_promotions"

    id = Column(Integer, primary_key=True, index=True)
    source_cluster_id = Column(
        Integer, ForeignKey("kubernetes_clusters.id", ondelete="SET NULL"), nullable=True
    )
    target_cluster_id = Column(
        Integer, ForeignKey("kubernetes_clusters.id", ondelete="SET NULL"), nullable=True
    )

    # Promotion metadata
    source_cluster_name = Column(String(255), nullable=False)
    target_cluster_name = Column(String(255), nullable=False)
    change_count = Column(Integer, nullable=False, default=0)
    changes_summary = Column(JSON, nullable=True)  # List of {action, kind, name, namespace}
    applied_by = Column(String(255), nullable=True)

    # Result
    success = Column(Boolean, default=True)
    error_message = Column(String(1000), nullable=True)
    applied_count = Column(Integer, default=0)
    failed_count = Column(Integer, default=0)
    skipped_count = Column(Integer, default=0)

    applied_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    # Relationships
    source_cluster = relationship(
        "KubernetesCluster", foreign_keys=[source_cluster_id]
    )
    target_cluster = relationship(
        "KubernetesCluster", foreign_keys=[target_cluster_id]
    )

    __table_args__ = (
        Index("idx_config_promotion_applied_at", "applied_at"),
        Index("idx_config_promotion_source", "source_cluster_id"),
        Index("idx_config_promotion_target", "target_cluster_id"),
    )
